actionpack/lib/action_dispatch/journey/router.rb

```diff
@@ -122,6 +122,7 @@ def find_routes req
 
         def match_head_routes(routes, req)
           head_routes = match_routes(routes, req)
+          head_routes.delete_if { |route| route.verb == // }
```

Comment:
The problem comes because we are explicitly rejecting catch-all routes when doing a HEAD request. Since a head request is a special case of a GET request, we don't want to allow anything that doesn't explicitly accept HEAD requests. However, after the initial pass-through of the routes, and nothing is found, we fall back to allow GET routes (and catch-all routes) to respond. As a result, if we put the rejection in `matches?` the fallback becomes more difficult.